parser_api/example.rb

```diff
@@ -0,0 +1,173 @@
+      model.eAllContents.each {|m| check_bad_word(m) }
+      @acceptor
+    end
+
+    # perform the bad_word check on one AST element
+    # (this is done using a polymorphic visitor)
+    #
+    def check_bad_word(o)
+      @@bad_word_visitor.visit_this_0(self, o)
+    end
+
+    protected
+
+    def badword_Object(o)
+      # ignore all not covered by an explicit badword_xxx method
+    end
+
+    # A bare word is a QualifiedName
+    #
+    def badword_QualifiedName(o)
+      if o.value == 'bigly'
+        acceptor.accept(Issues::INVALID_WORD, o, :text => o.value)
+      end
+    end
+  end
+
+  class MyFactory < Puppet::Pops::Validation::Factory
+    # Produces the checker to use
+    def checker(diagnostic_producer)
+      MyChecker.new(diagnostic_producer)
+    end
+
+    # Produces the label provider to use.
+    #
+    def label_provider
+      # We are dealing with AST, so the existing one will do fine.
+      # This is what translates objects into a meaningful description of what that thing is
+      #
+      Puppet::Pops::Model::ModelLabelProvider.new()
+    end
+
+    # Produces the severity producer to use. Here it is configured what severity issues have
+    # if they are not all errors. (If they are all errors this method is not needed at all).
+    #
+    def severity_producer
+      # Gets a default severity producer that is then configured below
+      p = super
+
+      # Configure each issue that should **not** be an error
+      #
+      p[Issues::INVALID_WORD]                 = :warning
+
+      # examples of what may be done here
+      # p[Issues::SOME_ISSUE]           = <some condition> ? :ignore : :warning
+      # p[Issues::A_DEPRECATION]        = :deprecation
+
+      # return the configured producer
+      p
+    end
+
+    # Allow simpler call when not caring about getting the actual acceptor
+    def diagnostic_producer(acceptor=nil)
+      acceptor.nil? ? super(Puppet::Pops::Validation::Acceptor.new) : super(acceptor)
+    end
+  end
+
+  # We create a diagnostic formatter that outputs the error with a simple predefined
+  # format for location, severity, and the message. This format is a typical output from
+  # something like a linter or compiler.
+  # (We do this because there is a bug in the DiagnosticFormatter's `format` method prior to
+  # Puppet 4.9.0. It could otherwise have been used directly.
+  #
+  class Formatter < Puppet::Pops::Validation::DiagnosticFormatter
+    def format(diagnostic)
+      "#{format_location(diagnostic)} #{format_severity(diagnostic)}#{format_message(diagnostic)}"
+    end
+  end
+end
+
+# -- Example usage of the new validator
+
+# Get a parser
+parser = Puppet::Pops::Parser::EvaluatingParser.singleton
+
+# parse without validation
+result = parser.parser.parse_string('$x = if 1 < 2 { smaller } else { bigly }', 'testing.pp')
+result = result.model
+
+# validate using the default validator and get hold of the acceptor containing the result
+acceptor = parser.validate(result)
```

Comment:
Perhaps make the `validate` method capable of validating the _result_ from `parse_string` rather than the model that it contains. That way, the _result_ becomes opaque (i.e. no need to disclose the somewhat unintuitive implementation detail `result = result.model` ).